software/performance_counters/performance_counters.c

```diff
@@ -0,0 +1,111 @@
+// See LICENSE for license details.
+
+// This demo shows how to use basic
+// RISC-V profiling counters, mcycle
+// (counts the number of processor cycles)
+// and minstret (counts the number of retired instructions). 
+// Note that both are writable as well.
+
+#include <stdio.h>
+
+// The CSR encodings are in this header.
+#include "encoding.h"
+
+// The mcycle counter is 64-bit counter, but since
+// Freedom E platforms use RV32, we must access it as
+// 2 32-bit registers. At 256MHz, the lower bits will
+// roll over approx. every 5 seconds, so we check for
+// rollover with this routine as suggested by the
+// RISC-V Priviledged Architecture Specification.
+
+#define rdmcycle(x)  {				       \
```

Comment:
well, it should be inlined, etc.... we don't really want it taking extra cycles to return stuff. 

But agreed that this can/should be moved to a shared library.